src/lib/support/BufBound.h

```diff
@@ -0,0 +1,137 @@
+/*
+ *
+ *    Copyright (c) 2020 Project CHIP Authors
+ *
+ *    Licensed under the Apache License, Version 2.0 (the "License");
+ *    you may not use this file except in compliance with the License.
+ *    You may obtain a copy of the License at
+ *
+ *        http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *    Unless required by applicable law or agreed to in writing, software
+ *    distributed under the License is distributed on an "AS IS" BASIS,
+ *    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *    See the License for the specific language governing permissions and
+ *    limitations under the License.
+ */
+
+/**
+ *  @file
+ *    basic double-ended queue, intended to be embedded as a member
+ *     of an object
+ */
+
+#ifndef CHIP_BUFBOUND_H
+#define CHIP_BUFBOUND_H
+
+#include <stdint.h>
+#include <string.h>
+
+namespace chip {
+
+/**
+ *  @class BufBound
+ *
+ *  simple buffer writer
+ */
+class BufBound
+{
+private:
+    uint8_t * mBuf;
+    size_t mLen;
+    size_t mWritten;
+
+public:
+    BufBound(uint8_t * buf, size_t len) : mBuf(buf), mLen(len), mWritten(0) {}
+
+    /*
+     * @brief append a single byte
+     */
+    size_t Put(uint8_t c)
+    {
+        if (mWritten < mLen)
+        {
+            mBuf[mWritten] = c;
+        }
+        return ++mWritten;
+    }
+
+    /*
+     * @brief append a null terminated string, exclude the null term
+     */
+    size_t Put(const char * s)
+    {
+        while (*s != 0)
+        {
+            Put(*s++);
+        }
+        return mWritten;
+    }
+
+    size_t PutLE(uint64_t x, size_t size)
+    {
+        while (size-- > 0)
+        {
+            uint8_t c = x & 0xff;
+            Put(c);
+            x >>= 8;
+        }
+        return mWritten;
+    }
+
+    /*
+     * @brief write integer x into the buffer, least significant byte first
+     */
+    size_t PutLE(uint64_t x) { return PutLE(x, sizeof(x)); }
```

Comment:
I would consider using different naming for different sizes besides overload.

Without it, to make sure the correct code is called, one should do `PutLE(static_cast<uint16_t>(1))` which seems more verbose than `Put_u16(1)` or similar. Something unqualified like `Put(1)` seems not specifiv enough and something like `Put(foo)` forces me to lookup and ensure that foo is the right variable size

Reply:
@rwalker-apple ?

Reply:
works for me.  wait one...

Reply:
@woody-apple done